pkg/client/transport.go

```diff
@@ -54,44 +54,84 @@ func (rt *bearerAuthRoundTripper) RoundTrip(req *http.Request) (*http.Response,
 	return rt.rt.RoundTrip(req)
 }
 
-func NewClientCertTLSTransport(certData, keyData, caData []byte) (*http.Transport, error) {
+// TLSConfigFor returns a tls.Config that will provide the transport level security defined
+// by the provided Config. Will return nil if no transport level security is requested.
+func TLSConfigFor(config *Config) (*tls.Config, error) {
+	hasCA := len(config.CAFile) > 0 || len(config.CAData) > 0
+	hasCert := len(config.CertFile) > 0 || len(config.CertData) > 0
+
+	if hasCA && config.Insecure {
+		return nil, fmt.Errorf("specifying a root certificates file with the insecure flag is not allowed")
+	}
+	var tlsConfig *tls.Config
+	switch {
+	case hasCert:
```

Comment:
The original code here (if I remember correctly) is - you told us you wanted insecure, but then you gave us a CA.  You can't do both, because CA without verify is pointless.

----- Original Message -----
> > @@ -54,44 +54,84 @@ func (rt *bearerAuthRoundTripper) RoundTrip(req
> > *http.Request) (*http.Response,
> >  	return rt.rt.RoundTrip(req)
> >  }
> >  
> > -func NewClientCertTLSTransport(certData, keyData, caData []byte)
> > (*http.Transport, error) {
> > +// TLSConfigFor returns a tls.Config that will provide the transport level
> > security defined
> > +// by the provided Config. Will return nil if no transport level security
> > is requested.
> > +func TLSConfigFor(config *Config) (*tls.Config, error) {
> > +	hasCA := len(config.CAFile) > 0 || len(config.CAData) > 0
> > +	hasCert := len(config.CertFile) > 0 || len(config.CertData) > 0
> > +
> > +	if hasCA && config.Insecure {
> > +		return nil, fmt.Errorf("specifying a root certificates file with the
> > insecure flag is not allowed")
> > +	}
> > +	var tlsConfig *tls.Config
> > +	switch {
> > +	case hasCert:
> 
> Shouldn't this case be `hasCert && hasCA`? `NewClientCertTLSConfig` requires
> both.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3959/files#r23881524